src/reducers/versions.spec.tsx

```diff
@@ -1832,6 +1836,33 @@ describe(__filename, () => {
       );
     });
 
+    it('appends scrollTo if requested', async () => {
+      const location = createFakeLocation({ pathname });
+      const history = createFakeHistory({ location });
+      const diffPosition = ScrollTarget.firstDiff;
+
+      const { dispatch, thunk } = thunkTester({
+        createThunk: () => {
+          return viewVersionFile({
+            selectedPath,
+            scrollTo: diffPosition,
+            versionId,
+          });
+        },
+        store: configureStore({ history }),
+      });
+
+      await thunk();
+
+      expect(dispatch).toHaveBeenCalledWith(
+        push({
+          ...location,
+          search: `?path=${selectedPath}&scrollTo=${diffPosition}`,
```

Comment:
Use [`expect.urlWithTheseParams()`](https://github.com/mozilla/addons-code-manager/blob/5ab911bda04346c6d5e9583f467992f9051b86ce/src/jest-matchers/index.tsx#L32) here if you can. Otherwise, you're relying on the order of object serialization which could be flaky one day.

Reply:
Good idea. I am having trouble figuring out the syntax for using it in this situation as we are not asserting about a property of an object that was passed into the call to `dispatch`, but rather about a property of an object that was passed into the call of `push`. I tried this:

```js
      expect(dispatch).toHaveBeenCalledWith(
        push(
          expect.objectContaining({
            hash: undefined,
            search: expect.urlWithTheseParams({
              path: selectedPath,
              scrollTo: diffPosition,
            }),
          }),
        ),
      );
```

Which yields this failure:

```
  ● /Users/bsilverberg/Documents/gitRepos/addons-code-manager/src/reducers/versions.spec.tsx › viewVersionFile › appends scrollTo if requested

    expect(jest.fn()).toHaveBeenCalledWith(expected)

    Expected mock function to have been called with:
      {"payload": {"args": [ObjectContaining {"hash": undefined, "search": urlWithTheseParams<[object Object]>}], "method": "push"}, "type": "@@router/CALL_HISTORY_METHOD"}
    as argument 1, but it was called with
      {"payload": {"args": [{"key": "some-key", "pathname": "/some/path/to/a/page", "search": "?path=some-path&scrollTo=firstDiff", "state": {}}], "method": "push"}, "type": "@@router/CALL_HISTORY_METHOD"}.

    Difference:

    - Expected
    + Received

      Object {
        "payload": Object {
          "args": Array [
    -       ObjectContaining {
    -         "hash": undefined,
    -         "search": urlWithTheseParams<[object Object]>,
    +       Object {
    +         "key": "some-key",
    +         "pathname": "/some/path/to/a/page",
    +         "search": "?path=some-path&scrollTo=firstDiff",
    +         "state": Object {},
            },
          ],
          "method": "push",
        },
        "type": "@@router/CALL_HISTORY_METHOD",
      }

      {"payload": {"args": [ObjectContaining {"hash": undefined, "search": urlWithTheseParams<[object Object]>}], "method": "push"}, "type": "@@router/CALL_HISTORY_METHOD"}
    as argument 1, but it was called with
      {"error": undefined, "meta": undefined, "payload": {"selectedPath": "some-path", "versionId": 123}, "type": "UPDATE_SELECTED_PATH"}.

    Difference:

    - Expected
    + Received

      Object {
    +   "error": undefined,
    +   "meta": undefined,
        "payload": Object {
    -     "args": Array [
    -       ObjectContaining {
    -         "hash": undefined,
    -         "search": urlWithTheseParams<[object Object]>,
    -       },
    -     ],
    -     "method": "push",
    +     "selectedPath": "some-path",
    +     "versionId": 123,
        },
    -   "type": "@@router/CALL_HISTORY_METHOD",
    +   "type": "UPDATE_SELECTED_PATH",
      }

      1863 |       );
      1864 |
    > 1865 |       expect(dispatch).toHaveBeenCalledWith(
           |                        ^
      1866 |         push(
      1867 |           expect.objectContaining({
      1868 |             hash: undefined,

      at Object.toHaveBeenCalledWith (src/reducers/versions.spec.tsx:1865:24)
```

I'm not sure why it's showing me two match failures, so maybe that's a hint as to what the problem is? Is it because of the expect inside an expect inside an expect?

@kumar303 Do you have any advice?

Reply:
You probably need to do it in two calls:

```tsx
      expect(dispatch).toHaveBeenCalledWith(
        push(
          expect.objectContaining({
            hash: undefined,
          }),
        ),
      );
      expect(dispatch.calls[0].search).toEqual(
            expect.urlWithTheseParams({
              path: selectedPath,
              scrollTo: diffPosition,
            }),
      );
```

... but maybe it would work like this?

```tsx
expect(dispatch).toHaveBeenCalledWith(
          expect.objectContaining({
            hash: undefined,
            search: expect.urlWithTheseParams({
              path: selectedPath,
              scrollTo: diffPosition,
            }),
          }),
      );
```

Reply:
It turns out that:

```js
      expect(dispatch).toHaveBeenCalledWith(
        push(
          expect.objectContaining({
            pathname,
            search: expect.urlWithTheseParams({
              path: selectedPath,
              scrollTo: diffPosition,
            }),
          }),
        ),
      );
```

works, which is what I had originally, except that the difference is that I didn't include `hash: undefined` in the `expect.objectContaining`. That was the problem. Even though `hash` was not present in the actual object, including `hash: undefined` in the `expect.objectContaining` was failing because, I presume, it was looking for `hash: undefined` to be present in the object. Looking at the test, there's actually no reason for it to be checking anything about `hash` in the call anyway, so I removed that and now it's working.